### PR TITLE
Add EnVAR functionality

### LINKED ISSUE #99: Implementation of Ensembe VAR
### Add a 3denvar case and layout, no LETKF for the first implementation:
See loose representation of the workflow in the figure below.

EnVAR

### Generate EnVAR yaml file on the fly
- [x] Create 3dhybenvar.yaml file with proper hash tags
- [x] Recursively add members to above yml => create member template
- [x] Create BUMP localization initialization application/yaml config

- [ ] Try not to forget to add yaml files to the soca-config repository

### PR DESCRIPTION
**Pull request by guillaumevernieres**
This work implements a functional EnVAR, perturbations are generated at every cycle using the "genpert.x" application instead of the letkf (next pr).
To run the test, it needs the branch feature/emc-envar of soca-config.

See issue #99 for more details.
closes #99

### CODE REVIEW

File: workflow/defaults/default_resources.yaml [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 godas_resource_table:
 #                ranks  ppn  wallclock              threads MB
-  r3dvar:       [  100,   20, !timedelta "01:00:00", 12,  "3072" ]
+  r3dvar:       [  80,   20, !timedelta "01:00:00", 12,  "3072" ]
```
Comment (flampouris) [inline workflow/defaults/default_resources.yaml]: Add one exclusive entry for the envar, and for now, we can set the r3dvar:       [  20,   20, ...].

File: workflow/layout/3denvar_only.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        <<: *shared_task_template
+        Trigger: !Depend  obs_prep
+        resources: !calc partition.resources.run_da_prep
+        config: [ base ]
+        J_JOB: JJOB_DA_PREP
+
+      da_prep_soca: !Task
+        <<: *shared_task_template
+        Trigger: !Depend  da_prep
+        resources: !calc partition.resources.run_3dvar
```
Comment (flampouris) [inline workflow/layout/3denvar_only.yaml]: The name of the resources, when updated default_resources.yaml

File: workflow/defaults/places.yaml [hunk 3]
```diff
@@ -14,23 +14,23 @@ default_places: &default_places
   DA_SLOT_LEN: 24
 # To be removed as soon as there is an ensemble yaml file
 # Number of ensemble members
-  NO_ENS_MBR: 5
+  NO_ENS_MBR: 40
```
Comment (flampouris) [inline workflow/defaults/places.yaml]: 5?

File: scripts/prep_3dvar_yaml.sh [hunk 4]
```diff
@@ -14,13 +14,13 @@
...
 # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # 
-#set -x
+# set -x
 
 echo '#=================================================================#'
 echo '#                    prep_3dvar_yaml.sh starts                    #'
 echo '#                                                                 #'
 
-while getopts "i:d:" opt; do
+while getopts "i:d:n:" opt; do
```
Comment (flampouris) [inline scripts/prep_3dvar_yaml.sh]: What's the n for?

File: scripts/prep_ensvariance_yaml.sh [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      n) Ne=("$OPTARG");;
+   esac
+done
+shift $((OPTIND -1))
+
+cd ${RUNDIR}
+
+echo $membertemplate
+for (( mbr=1; mbr<=${Ne} ; mbr++ )); do       
+    echo "I am member "$mbr
```
Comment (flampouris) [inline scripts/prep_ensvariance_yaml.sh]: Do we keep this?

File: scripts/prep_3denvar_yaml.sh [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      n) Ne=("$OPTARG");;
+   esac
+done
+shift $((OPTIND -1))
+
+cd ${RUNDIR}
+
+echo $membertemplate
+for (( mbr=1; mbr<=${Ne} ; mbr++ )); do
+    echo "I am member "$mbr
```
Comment (flampouris) [inline scripts/prep_3denvar_yaml.sh]: We can skip this,no?

File: jobs/JJOB_PREP_SOCA [hunk 7]
```diff
@@ -41,3 +37,11 @@ echo 'Initializing BUMP.'
 echo
 mkdir -p bump
 srun -n $NPE ${SOCA_EXEC}/soca_staticbinit.x ./yaml/static_SocaError_init.yml
+
+# Generate BUMP localization
+if [ ${NO_ENS_MBR}>1 ]; then
+    echo
+    echo 'Initializing BUMP Localization.'
+    echo
+    srun -n $NPE ${SOCA_EXEC}/soca_parameters.x ./yaml/parameters_bump_loc.yml
```
Comment (flampouris) [inline jobs/JJOB_PREP_SOCA:46]: What's the soca_paremeters for?

File: jobs/JJOB_DA_3DENVAR [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+# Unlink old restarts
+rm -f ./ANA/MOM.*.nc
+rm -f $PWD/ANA/cice_bkg.nc
+# Get MOM's restart
+mv -f RESTART/MOM.*.nc ./ANA/  # Because checkpoint dumps analysis in RESTART
+ln -s $PWD/Data/cic.socagodas.an.*.nc ./ANA/cice_bkg.nc
+
+cat <<EOF
+#============================================
+# Run mom6 only deterministic forecast
```
Comment (flampouris) [inline jobs/JJOB_DA_3DENVAR:58]: Shouldn't this be at the forecast job?

Comment (guillaumevernieres) [reply jobs/JJOB_DA_3DENVAR:58]: Yes, it should. But that need to be addressed in a different pr, the 3dvar is already implemented this way.

File: jobs/JJOB_DA_PREP [hunk 9]
```diff
@@ -98,12 +104,46 @@ ${ROOT_GODAS_DIR}/scripts/replace_KWRD_yaml.sh     \
...
+      -i $RUNCDATE/yaml/3dvar_godas.yml            \
       -d $RUNCDATE
 
+# Prep yaml for 3denvar 
+#----------------------------------
+${ROOT_GODAS_DIR}/scripts/prep_3denvar_yaml.sh    \
+      -i $RUNCDATE/yaml/3dhybenvar.yml            \
+      -m $RUNCDATE/yaml/envar_member.yml          \
+      -d $RUNCDATE                                \
+      -n ${NO_ENS_MBR}
```
Comment (flampouris) [inline jobs/JJOB_DA_PREP:135]: We use two different letters for the ${NO_ENS_MBR} -n and -v, let's keep only one.

Comment (guillaumevernieres) [reply jobs/JJOB_DA_PREP:135]: Ok, but:
-**i**nput
-**m**ember
-**d**ate
-**n**umber of member ... what does "v" stand for in the other script?

**Review by flampouris — CHANGES_REQUESTED** [review]
The 3dvar case fails. The log file is at /scratch2/NCEPDEV/marine/Stylianos.Flampouris/scrub/sandbox/test_3dvar_with3denvar/log/2011100112

File: jobs/JJOB_PREP_SOCA (hunk 7 above)
Comment (guillaumevernieres) [reply jobs/JJOB_PREP_SOCA:46]: Initialization of the localization.

**Comment by guillaumevernieres** [comment]
@flampouris : Addressed your comments and fixed the bug in the 3dvar

**Review by flampouris — CHANGES_REQUESTED** [review]
It went through.

**Review by flampouris — APPROVED** [review]

**Pull request merged**